abci/example/kvstore/kvstore.go

```diff
@@ -116,19 +158,54 @@ func (app *Application) HandleTx(tx []byte) *types.ResponseDeliverTx {
 	return &types.ResponseDeliverTx{Code: code.CodeTypeOK, Events: events}
 }
 
+func (app *Application) Close() error {
+	app.mu.Lock()
+	defer app.mu.Unlock()
+
+	return app.state.db.Close()
+}
+
 func (app *Application) FinalizeBlock(req types.RequestFinalizeBlock) types.ResponseFinalizeBlock {
-	txs := make([]*types.ResponseDeliverTx, len(req.Txs))
+	app.mu.Lock()
+	defer app.mu.Unlock()
+
+	// reset valset changes
+	app.ValUpdates = make([]types.ValidatorUpdate, 0)
+
+	// Punish validators who committed equivocation.
+	for _, ev := range req.ByzantineValidators {
+		if ev.Type == types.EvidenceType_DUPLICATE_VOTE {
+			addr := string(ev.Validator.Address)
+			if pubKey, ok := app.valAddrToPubKeyMap[addr]; ok {
+				app.updateValidator(types.ValidatorUpdate{
+					PubKey: pubKey,
+					Power:  ev.Validator.Power - 1,
+				})
+				app.logger.Info("Decreased val power by 1 because of the equivocation",
+					"val", addr)
+			} else {
+				app.logger.Error("Wanted to punish val, but can't find it",
```

Comment:
I see that this predates the change here, but I believe the correct thing to do here is crash.